File: 463/main.py
"""
You are given a map in form of a two-dimensional integer grid where 1 represents land and 0 represents water. Grid cells are connected horizontally/vertically (not diagonally). The grid is completely surrounded by water, and there is exactly one island (i.e., one or more connected land cells). The island doesn't have "lakes" (water inside that isn't connected to the water around the island). One cell is a square with side length 1. The grid is rectangular, width and height don't exceed 100. Determine the perimeter of the island.
"""

class Solution(object):
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        total_sides = 4 * sum(list(map(sum, grid)))
        repeated_sides = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                curr = grid[i][j]
                if curr == 1:
                    if i != 0 and grid[i-1][j] == 1:
                        repeated_sides += 1
                    if j != 0 and grid[i][j-1] == 1:
                        repeated_sides += 1
        return total_sides - 2 * repeated_sides



ans = Solution()
print(ans.islandPerimeter(
[[0,1,0,0],
 [1,1,1,0],
 [0,1,0,0],
 [1,1,0,0]]))

File: 463/main.go
package main

import "fmt"

func main() {
	matrix := [][]int{
		[]int{0, 1, 0, 0},
		[]int{1, 1, 1, 0},
		[]int{0, 1, 0, 0},
		[]int{1, 1, 0, 0},
	}
	fmt.Println(islandPerimeter(matrix))
}

func islandPerimeter(grid [][]int) int {
	totalSides := 0
	repeatedSides := 0
	for i, row := range grid {
		for j, cell := range row {
			if cell == 1 {
				if i != 0 && grid[i-1][j] == 1 {
					repeatedSides++
				}
				if j != 0 && grid[i][j-1] == 1 {
					repeatedSides++
				}
			}
			totalSides += cell * 4
		}
	}
	return totalSides - 2*repeatedSides
}
